pkg/util/errutil/errors.go

```diff
@@ -1,9 +1,184 @@
+// returned for the error, and the msgID is passed to the caller
+// to serve as the base for user facing error messages.
+//
+// msgID should be structured as component.error-brief, for example
+//   login.failed-authentication
+//   dashboards.validation-error
+//   dashboards.uid-already-exists
+func NewBase(reason StatusReason, msgID string, opts ...BaseOpt) Base {
+	b := Base{
+		reason:    reason,
+		messageID: msgID,
+		logLevel:  reason.Status().LogLevel(),
+	}
+
+	for _, opt := range opts {
+		b = opt(b)
+	}
+
+	return b
+}
+
+type BaseOpt func(Base) Base
+
+// WithLogLevel sets a custom log level for all errors instantiated from
+// this Base.
+//
+// Used as a functional option to NewBase.
+func WithLogLevel(lvl LogLevel) BaseOpt {
+	return func(b Base) Base {
+		b.logLevel = lvl
+		return b
+	}
+}
+
+// WithPublicMessage sets the default public message that will be used
+// for errors based on this Base.
+//
+// Used as a functional option to NewBase.
+func WithPublicMessage(message string) BaseOpt {
+	return func(b Base) Base {
+		b.publicMessage = message
+		return b
+	}
+}
+
+// Errorf creates a new Error with the Reason and MessageID from
+// Base, and Message and Underlying will be populated using
+// the rules of fmt.Errorf.
+func (b Base) Errorf(format string, args ...interface{}) Error {
+	err := fmt.Errorf(format, args...)
+
+	return Error{
+		Reason:        b.reason,
+		LogMessage:    err.Error(),
+		PublicMessage: b.publicMessage,
+		MessageID:     b.messageID,
+		Underlying:    errors.Unwrap(err),
+		LogLevel:      b.logLevel,
+	}
+}
+
+// Error is the error type for errors within Grafana, extending
+// the Go error type with Grafana specific metadata to reduce
+// boilerplate error handling for status codes and internationalization
+// support.
+//
+// Error implements Unwrap and Is to natively support Go 1.13 style
+// errors as described in https://go.dev/blog/go1.13-errors .
+type Error struct {
+	Reason        StatusReason
+	MessageID     string
+	LogMessage    string
+	Underlying    error
+	PublicMessage string
+	PublicPayload map[string]interface{}
+	LogLevel      LogLevel
+}
+
+// MarshalJSON returns an error, we do not want raw Error:s being
+// marshaled into JSON.
+//
+// Use Public to convert the Error into a PublicError which can be
+// marshaled. This is not done automatically, as that conversion is
+// lossy.
+func (e Error) MarshalJSON() ([]byte, error) {
+	return nil, fmt.Errorf("errutil.Error cannot be directly marshaled into JSON")
+}
+
+// Error implements the error interface.
+func (e Error) Error() string {
+	return fmt.Sprintf("[%s] %s", e.MessageID, e.LogMessage)
+}
+
+// Log writes the error to a logger based on Error.LogLevel.
+//
+// If the logger is nil, this method is a no-op.
+func (e Error) Log(logger LogInterface) {
+	if logger == nil {
+		return
+	}
+
+	e.LogLevel.LogFunc(logger)(
+		e.MessageID,
+		"err", e.LogMessage,
+		"reason", e.Reason,
```

Comment:
We may need to include also the `traceID`

Reply:
Oh, yeah, I forgot about that. 😓
I was thinking about that a bit in the beginning, but I couldn't squeeze it into the API in a nice way. I figured that maybe that'd come later, not really. I'll circle around and try to do that.

Reply:
It's now included, I figured out how to reuse the logger in an ~ok way.